holoviews/core/data.py

```diff
@@ -1308,9 +1334,288 @@ def aggregate(cls, columns, kdims, function, **kwargs):
+                                 'shape do not match. Expected shape %s, '
+                                 'actual shape: %s' % (expected, vdim, shape))
+        return data, kdims, vdims
+
+
+    @classmethod
+    def validate(cls, columns):
+        DataColumns.validate(columns)
+
+
+    @classmethod
+    def dimension_type(cls, columns, dim):
+        if dim in columns.kdims:
+            arr = columns.data[dim.name]
+        elif dim in columns.vdims:
+            arr = columns.data[dim.name]
+        else:
+            return None
+        return arr.dtype.type
+
+
+    @classmethod
+    def shape(cls, columns):
+        return cls.length(columns), len(columns.dimensions()),
+
+
+    @classmethod
+    def length(cls, columns):
+        return np.product([len(columns.data[d.name]) for d in columns.kdims])
+
+
+    @classmethod
+    def values(cls, columns, dim, expanded=True, flat=True):
+        if dim in columns.kdims:
+            if not expanded:
+                return columns.data[dim]
+            prod = util.cartesian_product([columns.data[d.name] for d in columns.kdims])
+            idx = columns.get_dimension_index(dim)
+            return prod[:, idx]
+        else:
+            dim = columns.get_dimension(dim)
+            values = columns.data.get(dim.name)
+            return values.flatten() if flat else values
+
+
+    @classmethod
+    def groupby(cls, columns, dim_names, container_type, group_type, **kwargs):
+        # Get dimensions information
+        dimensions = [columns.get_dimension(d) for d in dim_names]
+        kdims = [kdim for kdim in columns.kdims if kdim not in dimensions]
+
+        # Update the kwargs appropriately for Element group types
+        group_kwargs = {}
+        group_type = dict if group_type == 'raw' else group_type
+        if issubclass(group_type, Element):
+            group_kwargs.update(util.get_param_values(columns))
+            group_kwargs['kdims'] = kdims
+        group_kwargs.update(kwargs)
+
+        # Find all the keys along supplied dimensions
+        keys = [columns.data[d.name] for d in dimensions]
+
+        # Iterate over the unique entries applying selection masks
+        grouped_data = []
+        for unique_key in util.cartesian_product(keys):
+            group_data = cls.select(columns, **dict(zip(dim_names, unique_key)))
+            for vdim in columns.vdims:
+                group_data[vdim.name] = np.squeeze(group_data[vdim.name])
+            group_data = group_type(group_data, **group_kwargs)
+            grouped_data.append((tuple(unique_key), group_data))
+
+        if issubclass(container_type, NdMapping):
+            with item_check(False), sorted_context(False):
+                return container_type(grouped_data, kdims=dimensions)
+        else:
+            return container_type(grouped_data)
+
+
+    @classmethod
+    def coord_mask(cls, columns, arr, ind):
+        if isinstance(ind, tuple):
```

Comment:
I've renamed it to key_select_mask because that's what made most sense to me. I'm still passing through the values array because it avoids getting it multiple times and I don't want to remove columns as an argument because it would be the only inconsistent method in that regard.